[GAWB-2021] Rawls admin endpoint: what's on the workflow queue?

rtitle opened the pull request:
See https://broadinstitute.atlassian.net/browse/GAWB-2021

Added `/api/admin/submissions/queueStatusByUser` endpoint for listing workflow counts by status and by user.

Unit tests pass. I tested locally against the dev DB, and got the following results:
```
{
  "statuses": {
    "Aborting": 937,
    "Submitted": 96
  },
  "users": {
    "joel.broad.dev@gmail.com": {
      "statuses": {
        "Submitted": 1
      }
    },
    "abaumann.firecloud@gmail.com": {
      "statuses": {
        "Aborting": 937
      }
    },
    "test.firec@gmail.com": {
      "statuses": {
        "Submitted": 21
      }
    },
    "obamacloud@gmail.com": {
      "statuses": {
        "Submitted": 74
      }
    }
  },
  "maxActiveWorkflowsTotal": 20000,
  "maxActiveWorkflowsPerUser": 2000
}
```

Here is the new database query:
```
rawls [DEBUG] [18:41:46.278] [slick.db-86] slick.jdbc.JdbcBackend.statement - Preparing statement: 
select x2.`SUBJECT_ID`, x2.`EMAIL`, x3.`ID`, x3.`EXTERNAL_ID`, x3.`SUBMISSION_ID`, x3.`STATUS`, x3.`STATUS_LAST_CHANGED`, x3.`ENTITY_ID`, x3.`record_version`, x3.`EXEC_SERVICE_KEY` from `WORKFLOW` x3, `SUBMISSION` x4, `USER` x2 where (x4.`SUBMITTER` = x2.`SUBJECT_ID`) and ((x3.`STATUS` in (?, ?, ?, ?, ?)) and (x3.`SUBMISSION_ID` = x4.`ID`))
rawls [DEBUG] [18:41:46.561] [slick.db-86] slick.jdbc.JdbcBackend.benchmark - Execution of prepared statement took 283ms
```
And its explain plan (it seems reasonable):
![image](https://cloud.githubusercontent.com/assets/5368863/25759850/369cb16e-31a2-11e7-81e6-cf95178a4e78.png)

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
  - _RT: see https://github.com/broadinstitute/firecloud-admin-cli/pull/37_
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

rtitle committed:
[GAWB-2021] Rawls admin endpoint: what's on the workflow queue?

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiService.scala:
```diff
@@ -275,6 +275,13 @@ trait AdminApiService extends HttpService with PerRequestCreator with UserInfoDi
         }
       }
+    } ~
+    path("admin" / "submissions" / "queueStatusByUser") {
```
I wasn't quite sure about the endpoint name -- if anyone has other suggestions, let me know.

jmthibault79 replied:
Seems reasonable to me.  Please move this up to group with the other `submissions` endpoints.

rtitle replied:
Will do.

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:
```diff
@@ -305,6 +313,34 @@ class ExecutionJsonSupport extends JsonSupport {
 
   implicit val WorkflowQueueStatusResponseFormat = jsonFormat3(WorkflowQueueStatusResponse)
+
+  implicit object WorkflowQueueStatusByUserResponseFormat extends RootJsonFormat[WorkflowQueueStatusByUserResponse] {
```
Originally this was a one-liner:
```
implicit val WorkflowQueueStatusByUserResponseFormat = jsonFormat4(WorkflowQueueStatusByUserResponse)
```
...which worked, but I needed a custom formatter to conform to the JSON schema in [GAWB-2021](https://broadinstitute.atlassian.net/browse/GAWB-2021). I think this JSON format is more clear anyway.

jmthibault79 replied:
Could you add a test for the JSON formatting?  Could be as simple as a round-trip JSON string comparison.

rtitle committed:
Fix YAML syntax error

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -285,6 +287,20 @@ trait WorkflowComponent {
+        user <- rawlsUserQuery if submission.submitterId === user.userSubjectId
+      } yield (user, workflow)).result
+
+      // Combine results into a nested map using Scalaz semigroups for aggregation.
```
Since we don't currently have any other uses of semigroups in Workbench, I'm concerned that other developers (including me) won't be able to reason about what's going on here, from lack of familiarity.

Could you accomplish this in another way, or add a detailed comment about how this works?

rtitle replied:
Heh, I was trying to gauge people's reactions by importing scalaz (I noticed it's already a dependency!).

I feel like a lot of FP ideas have confusing names, but are actually relatively simple concepts. In particular a semigroup abstracts _combining_ things of the same type. Scalaz provides implicit semigroups for several different types, which can all be invoked with the `|+|` operator. 

For example, the semigroup for `Int` is addition:
```
scala> import scalaz._, Scalaz._
import scalaz._
import Scalaz._

scala> 1 |+| 2
res0: Int = 3
```
The semigroup for `List` is concatenation:
```
scala> List(1,2,3) |+| List(4,5,6)
res1: List[Int] = List(1, 2, 3, 4, 5, 6)
```
The semigroup for `Map` groups elements of the same key, and applies the semigroup for the value type:
```
scala> Map('a' -> 1, 'b' -> 2) |+| Map('a' -> 42, 'c' -> 3)
res2: scala.collection.immutable.Map[Char,Int] = Map(a -> 43, c -> 3, b -> 2)
```

You get the idea. So in this case we're using the semigroups for `Map` and `Int` to reduce a list of tuples into a `Map[String, Map[String, Int]]`. 

I could write it without scalaz using standard library functions, but I would argue it would make the code _more_ complicated. :) That said, I can give it a shot.

Also, as I was discussing with @helgridly, I'd be happy to give workbench folks an overview of some FP concepts, like semigroups, monoids, functors, monads, etc if people are interested. :)

jmthibault79 replied:
Thanks for the explanation.  There are definitely other places in Rawls where we do this, and it might be good to convert them to semigroups once we're more comfortable.  But for now, I would like a comment along the lines of "this operator combines maps' keys and sums their values"

A general overview of these concepts would be excellent.  We're pretty comfortable with monads (or at least map/flatMap) but not the rest of the family.  Red Team (Cromwell) does more Scalaz.  As we grow closer organizationally maybe they will have more of an influence on us...

helgridly replied:
Jeff has been banging the "we can afford to FP a little harder" drum for a while and I think this might be the thing that pushes us over the edge. I'm running it by the other TLs now; watch this space, but my money's on:

1) general consensus that this is worth doing;
2) a preference for cats over scalaz; and
3) a tech talk or two to properly introduce a couple of these concepts to the team.

rtitle replied:
Yay! I can port this to cats (I think it's the same syntax) and add more explanatory comments if people are ok with it.

jmthibault79 commented:
Looks good!  Could you add a few more tests which start with adding workflows, and also add some numbers for expected counts to show that we're not just comparing 0 to 0.

A nice way to check Swagger syntax is by pasting it in here: http://bigstickcarpet.com/swagger-parser/www/index.html

I like the Explain Plan diagram - could you show me how you made it sometime?  :)

rtitle commented:
Thanks @jmthibault79! I'll add some tests.

Thanks for the swagger tip, it took me a while to find the syntax issue. :(

Sure, I just ran the query in MySQL Workbench and scrolled down to "Execution Plan".

jmthibault79 commented:
Oh cool I should really try out MySQL Workbench!

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -285,6 +287,20 @@ trait WorkflowComponent {
+        workflow <- findQueuedAndRunningWorkflows
+        submission <- submissionQuery if workflow.submissionId === submission.id
+        user <- rawlsUserQuery if submission.submitterId === user.userSubjectId
+      } yield (user, workflow)).result
```
So this returns all the workflows from the database, and then you count them in Scala. That's a _lot_ of rows. Why not group by and count in the database?

helgridly replied:
In half-assed pseudo-SQL: `count(*) join workflows, submission, users, group by (workflow.status, user.id)`.

rtitle replied:
That's a good point.. it might be more efficient to do that in the database. Let me try that out next week.

danbills commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -285,6 +287,20 @@ trait WorkflowComponent {
+      userWorkflows.map(_.foldLeft(Map.empty[String, Map[String, Int]]) { case (result, (user, workflow)) =>
+        result |+| Map(user.userEmail -> Map(workflow.status -> 1))
+      })
+    }
```
a foldMap would clean this up a little.

```scala
      userWorkflows.map(_.toList.foldMap { case (user, workflow) =>
        Map(user.userEmail -> Map(workflow.status -> 1))
      })
```

rtitle replied:
Cool, thanks for the tip. I'm actually refactoring this to do the aggregation in the database via slick instead of using scalaz. Foldable/foldMap is neat though. :)

rtitle replied:
Btw I came across this, seemed kinda interesting re. slick and cats: https://github.com/RMSone/slick-cats

geoffjentry replied:
Not being published is kind of a pita though :( but yeah

rtitle committed:
Code review comments:
- Add test cases which create workflows and validate expected counts
- Updated WorkflowComponent to do grouping in the database instead of in Scala code

rtitle commented:
@helgridly @jmthibault79 pushed commit with code review feedback:

1. Added test cases which create some workflows/users and validate expected counts.
2. Added `ExecutionModelSpec` for testing JSON serialization/deserialization.
3. Updated `WorkflowComponent` to do the grouping in the database instead of in Scala. The new query is something like:
```
select x2.`EMAIL`, x3.`STATUS`, count(1) 
from `WORKFLOW` x3, `SUBMISSION` x4, `USER` x2 
where (x4.`SUBMITTER` = x2.`SUBJECT_ID`) 
and ((x3.`STATUS` in ('Queued', 'Launching', 'Submitted', 'Running', 'Aborting')) 
and (x3.`SUBMISSION_ID` = x4.`ID`)) 
group by x2.`EMAIL`, x3.`STATUS`
```
And the explain plan shows the group:
![image](https://cloud.githubusercontent.com/assets/5368863/25826128/9874eb5e-3413-11e7-8d5f-f3b744a51c14.png)

rtitle commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/RawlsEnumeration.scala:
```diff
@@ -1,6 +1,6 @@
 package org.broadinstitute.dsde.rawls.model
 
-trait RawlsEnumeration[T <: RawlsEnumeration[T]] { self: T =>
+trait RawlsEnumeration[T <: RawlsEnumeration[T]] extends Product with Serializable { self: T =>
```
Btw, https://github.com/lloydmeta/enumeratum is a nice Scala enum library.

jmthibault79 replied:
That does look nice.

Why was this change needed?

rtitle replied:
Good question, type inference! Basically all case classes automatically extend Product and Serializable, so to avoid it showing up in inferred types (i.e. when using case classes with parent traits in collections) you have to add it to the base trait.

I could have also added it to the WorkflowStatus trait, but I'm kind of assuming that all concrete implementations of RawlsEnumeration will be case classes/objects.

This SO article helps explain it a little better: http://stackoverflow.com/questions/36525804/scala-case-class-extending-product-with-serializable

geoffjentry replied:
Another explanation, including why one should use final as well: http://underscore.io/blog/posts/2015/06/04/more-on-sealed.html

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala:
```diff
@@ -1153,13 +1173,19 @@ class AdminApiServiceSpec extends ApiServiceSpec {
           .filterKeys((WorkflowStatuses.queuedStatuses ++ WorkflowStatuses.runningStatuses).map(_.toString).contains)
           .mapValues(_.size)
-        val expected = WorkflowQueueStatusByUserResponse(
+        val userOwnerWorkflows = (testData.userOwner.userEmail.value ->
```
Some nitpicks: please put the `testUserWorkflows` first and add a comment for the `userOwnerWorkflows` describing that that these are equal to all workflows in the systems besides those of the test user.  Renaming `statusCounts` to something like `testUserStatusCounts` would help to clarify that too.

Since this test would now break if we ever added a third user's workflows to the Default Test Database, this is a good candidate for using `withConstantTestDataApiServices` instead.  We try never to change that one.

rtitle replied:
Makes sense, will fix. Good tip about withConstantTestDataApiServices.

jmthibault79 commented:
The new tests are great and the DB logic makes sense.  I'll have to take your word on the `cats` stuff until I understand it.  👍

rtitle committed:
Unit test improvements for clarity

rtitle commented:
Thanks @jmthibault79! I pushed the unit test improvements.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:
```diff
@@ -214,6 +214,14 @@ case class WorkflowQueueStatusResponse
 
+case class WorkflowQueueStatusByUserResponse
+(
+  statuses: Map[String, Int],
```
Can we type alias this as `StatusCounts`, and _possibly_ do something similar for `users`? `users.convertTo[Map[String, Map[String, Map[String, Int]]]]` on 336 below is a bit of a mindbender.

rtitle replied:
Sure, can add type aliases for `statuses` and `users`.

helgridly commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponentSpec.scala:
```diff
@@ -251,6 +255,69 @@ class SubmissionComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
+    val testUserEmail = "testUser"
+    val testUserStatusCounts = Map(WorkflowStatuses.Submitted -> 1, WorkflowStatuses.Running -> 10, WorkflowStatuses.Aborting -> 100)
+    withWorkspaceContext(testData.workspace) { ctx =>
+      val testUser = RawlsUser(UserInfo(testUserEmail, OAuth2BearerToken("token"), 123, "123456789876543212346"))
```
Isn't this already `testData.userWriter`? I think this user has already been created and saved.

rtitle replied:
Whoops yeah, copy-paste. Will make the id unique.

helgridly commented on core/src/test/scala/org/broadinstitute/dsde/rawls/model/ExecutionModelSpec.scala:
```diff
@@ -0,0 +1,69 @@
+    // Verify round trip JSON serialization/deserialization
+    val testJson = testResponse.toJson
+    testJson should equal (expectedJson)
+    testJson.convertTo[WorkflowQueueStatusByUserResponse] should equal (testResponse)
```
Nice.

helgridly commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala:
```diff
@@ -1140,4 +1140,56 @@ class AdminApiServiceSpec extends ApiServiceSpec {
+    val testUserEmail = "testUser"
+    val testUserStatusCounts = Map(WorkflowStatuses.Submitted -> 1, WorkflowStatuses.Running -> 10, WorkflowStatuses.Aborting -> 100)
+    withWorkspaceContext(constantData.workspace) { ctx =>
+      val testUser = RawlsUser(UserInfo(testUserEmail, OAuth2BearerToken("token"), 123, "123456789876543212346"))
```
ditto above

helgridly commented:
A couple more nitpicks (sorry) but after those you're good to go. 👍

helgridly commented:
Do you need an entirely new user? Or can you reuse one?

rtitle committed:
Code review comments: add type aliases for clarity and use unique user ids in tests

rtitle commented:
I wanted a new user so I could guarantee their expected workflow counts. An existing user might have workflows not controlled by this test.

rtitle commented:
@helgridly pushed 1 more commit ^, let me know if approved.

helgridly commented:
do it
